File: books/competitive_programming/6-strings/secret_word.cpp
//ACCEPTED
//TAGS: KMP
#include <bits/stdc++.h>
using namespace std;

vector <int> back;
char s[2000001];
int len;

int kmp_preprocess(){
    int i = 0, j = -1, ans = 0;
    back.reserve(len+1); back[0] = -1;
    while (i < len){
        while (j >= 0 && s[i] != s[j]) j = back[j];
        i++; j++; back[i] = j;
        if (i > len/2) ans = max(ans, j);
    }
    return ans;
}

int main() {
    int t, k;
    string w, r;
    cin >> t;
    while (t--) {
        cin >> s;
        len = strlen(s);
        int i = len+1;
        s[len] = '$';
        for (int j = len-1; j >= 0; j--) {
            s[i] = s[j];
            i++;
        }
        len *= 2;
        len++;
        k = kmp_preprocess();
        for (int l = k-1; l >= 0; l--) {
            cout << s[l];
        }
        cout << "\n";
    }
}

File: books/competitive_programming/6-strings/secret_word.py
def kmp_preprocess(s):
    i = 0
    j = -1
    b = [0 for _ in range(len(s)+1)]
    b[0] = -1
    ans = 0
    while i < len(s):
        while j >= 0 and s[i] != s[j]: j = b[j]
        i += 1
        j += 1
        b[i] = j
        ans = max(ans, j)
        if i == len(s)//2: ans = 0
    print (b)
    return ans

t = int(input())
while t:
    t -= 1
    s = input().strip()
    print(s[0:kmp_preprocess(s+s[::-1])][::-1])


